design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,240 @@
+# Rsyslog - Proposal for New Input and Output Role Configuration
+
+## Background
+
+The logging role defines `rsyslog_default` variable to deploy the original, all-in-one rsyslog configuration file rsyslog.conf.
+It configures to get inputs from `imjournal` and output to the local files in /var/log.
+It was originally added to provide "noop" to the logging role.
+Then, it was utilized to support for the local system logs to store into the local files combining with the openshift logging in the ovirt logging.
+In the linux-system-roles review process, some critical questions on `rsyslog_default` were raised, such as the original intention - providing "noop" is not the style ansible recommends and no other linux-system-roles do not nor will not have such a default variable.
+Considering the inputs, the functionality that `rsyslog_default` offers has to be achieved by the logging role itself instead of copying the hardcoded rsyslog.conf to the target host.
+
+This proposal will discuss about these items:
+- how to replace the `rsyslog_default` variable with the logging role input/output role configuration.
+- how to support more generic way to connect input roles and output roles.
+
+Please note that this github commit starting this doc implements the first cut of the proposal including:
+- input roles - basics input role (`imjournal`, `imuxsock`, `imudp`, `imptcp`) and files input role (`imfile`), and
+- output roles - files output role (`omfile`), forwards output role (`omfwd`) and elasticsearch role (`omelasticsearch`).
+Leftover tasks, other roles to be changed and eliminating `rsyslog_default`, are discussed in [To Do's](#to-dos).
+
+## Proposals
+
+### Output roles
+Each output action is in the ruleset which name is the logging_outputs name.
+
+Elasticsearch, Files and Forwards output configuration files in /etc/rsyslog.d:
+```
+  10-output-modules.conf
+  30-output-elasticsearch.conf
+  30-output-files.conf
+  30-output-forwards.conf
+```
+With configuring the outputs (in this example, `files`) with a unique name (e.g., "files_output1", etc.), it will generate an rsyslog configuraion in the ruleset which name is the unique name.
+```
+ruleset(name="files_output1") {
+  *.info;authpriv.none;auth.none;cron.none;mail.none /var/log/messages
+}
+ruleset(name="files_output2") {
+  *.emerg :omusrmsg:*
+}
```

Comment:
would it work to put each of those into a separate configuration file?
I.e. have e.g. `30-output-files_output1.conf` containing
```yaml
ruleset(name="files_output1") {
  *.info;authpriv.none;auth.none;cron.none;mail.none /var/log/messages
}
```
and `30-output-files_output2.conf` containing
```yaml
ruleset(name="files_output2") {
  *.emerg :omusrmsg:*
}
```
This way, each unique `name` corresponds to a separate configuration file. And it is easy to implement `state: absent` by simply removing the corresponding file. But I am not an rsyslog expert and I don't know whether rsyslog would allow this.

Reply:
No, there is only one config file per output.  I think it'd be quite difficult to support the multiple files per output due to our usage of the jinja templates...

Reply:
So it is not a rsyslog limitation, just a limitation of the role? Why does our use of jinja2 templates make it difficult? It would be so logical to have one file for each `name` ...

Reply:
From rsyslog POV you can have as many files for output as you want, either put more PRI-FAC simplified rules after each other in the ruleset (the way its done in default config), or use traditional action() statements after the prio/facility specifier if you want to specify more options for the file output. 
You can have it inside one config file, or nest another includes within the ruleset and have it separate each file in its own config file. The second way also means you can chain the outputs, if that is something you would like.

As for the prior question, if you delete the config file defining a ruleset you are calling elsewhere rsyslog will run and sort-of ignore the call, but there will be an error upon startup as rsyslog expects everything that is used to also be defined, so I would not recommend this approach.

Reply:
Thank you, @pcahyna, @richm, @jvymazal, for your inputs.  Ok, rethinking about the implementation, it may not that hard to generate as many config files per ansible input config vars from customers (but it'll be in the separate pr :).